### ISSUE TITLE
Lazy datasets are not compatible with distributed training

### ISSUE THREAD

**Issue by JohnGiorgi** [allenai/allennlp#3924]
**Describe the bug**

I ran into an error when trying to train my model in a distributed setting using `lazy=True`

```bash
-- Process 1 terminated with the following error:
Traceback (most recent call last):
  File "/home/johnmg/t2t/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 19, in _wrap
    fn(i, *args)
  File "/scratch/johnmg/t2t/allennlp/allennlp/commands/train.py", line 464, in _train_worker
    metrics = train_loop.run()
  File "/scratch/johnmg/t2t/allennlp/allennlp/commands/train.py", line 523, in run
    return self.trainer.train()
  File "/scratch/johnmg/t2t/allennlp/allennlp/training/trainer.py", line 646, in train
    train_metrics = self._train_epoch(epoch)
  File "/scratch/johnmg/t2t/allennlp/allennlp/training/trainer.py", line 364, in _train_epoch
    batch_generator = iter(self.data_loader)
  File "/home/johnmg/t2t/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 279, in __iter__
    return _MultiProcessingDataLoaderIter(self)
  File "/home/johnmg/t2t/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 719, in __init__
    w.start()
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/process.py", line 112, in start
    self._popen = self._Popen(self)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/context.py", line 223, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'DatasetReader.read.<locals>.<lambda>'
```

and traced it down to this block in `DatasetReader`

https://github.com/allenai/allennlp/blob/bdb29a831ed68cb948b18b42fa61646b9ec11bf8/allennlp/data/dataset_readers/dataset_reader.py#L177-L183

My understanding is that in the distributed setting, the `DataLoader` must be pickled before being copied to each replica, and this pickling fails because of the `lambda` (I may be getting some of the details wrong here, but that's the gist).

I was able to produce a minimal example below that triggers the error and seems to suggest this is in fact the case.

**To Reproduce**

```python
import json
import pickle

from allennlp.data.dataset_readers import TextClassificationJsonReader

# Create a dummy dataset
data = {"text": "blah blah", "label": 0}
with open("data.jsonl", "w") as f: json.dump(data, f)

# Create a reader and read some data
reader = TextClassificationJsonReader(lazy=True)
dataset = reader.read("data.jsonl")

# Try to pickle the resulting dataset
with open("tmp.p", "wb") as f: pickle.dump(dataset, f)
```

Which gives you

```bash
AttributeError                            Traceback (most recent call last)
<ipython-input-22-59d848162c6c> in <module>
     13
     14 # Try to pickle the resulting dataset
---> 15 with open("tmp.p", "wb") as f: pickle.dump(dataset, f)

AttributeError: Can't pickle local object 'DatasetReader.read.<locals>.<lambda>'
```

the same error I reached when trying to train my own model and dataset reader.

I can avoid this error in the minimal example by importing dill as pickle (as per [this SO post](https://stackoverflow.com/questions/25348532/can-python-pickle-lambda-functions))

```python
import dill as pickle

# No error, (seemingly) pickles succesfully
with open("tmp.p", "wb") as f: pickle.dump(dataset, f) 
```

To be perfectly honest, I don't know why this works but am adding it here in case it is useful.

**Expected behavior**

I expect a model with a dataset reader for which `lazy=True` to train without error in the distributed case.

**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.7.4
 - AllenNLP version: "I installed from master"
 - PyTorch version: 1.4.0

**Additional context**

I was using my own custom dataset reader when I first bumped into the error, not `TextClassificationJsonReader`, so this is not just some issue with `TextClassificationJsonReader`. I only used it for the simplified example.

**Comment by matt-gardner** [comment]
This one (along with #3966) needs some serious design work and a bunch of new code, so I'm giving it a "week" label.

**Comment by JohnGiorgi** [comment]
Hate to be annoying, but is there any progress on this? Unfortunately it totally prevents training on multiple GPUs when `lazy=True`. I have to use a lazy dataset reader because of the size of my training data, but this also means I need multiple GPUs to train in a reasonable amount of time :/

**Comment by matt-gardner** [comment]
So, the immediate problem is being able to pickle the dataset, right?  It looks like there are two possibilities here: (1) try to get the multiprocessing code to use dill somehow (not sure how that works), or (2) change `_LazyInstances` to not take a `lambda`.  It looks like it might work if we just separate the one `Callable` argument [here](https://github.com/allenai/allennlp/blob/bdb29a831ed68cb948b18b42fa61646b9ec11bf8/allennlp/data/dataset_readers/dataset_reader.py#L44) into two arguments: a reference to the `_read` function (which should be pickleable) and the file path.  Then `_LazyInstances` just calls the function whenever it needs to, instead of using the `lambda`.  Does this second option make sense?  Can you try it and see if it solves your problem?  If so, then that's probably sufficient to close this issue.

**Comment by matt-gardner** [comment]
I had thought that this would require serious design work, but it might actually be solvable much more easily than I thought.  I'm not really clear on how the data loader interacts with these iterable datasets in reasonable ways, but if this workaround works then it should be good enough.

**Comment by JohnGiorgi** [comment]
@matt-gardner Okay great, that sounds like a reasonable solution to me. I just tried it, and while it does solve the issue, I run into a new issue:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
ModuleNotFoundError: No module named 't2t'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/multiprocessing/spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
ModuleNotFoundError: No module named 't2t'
```

even though I called `allennlp train` with `--include-package t2t` (and it works in the non-distributed case). I will spend some time debugging this but figured I would mention in first in case you spot an obvious issue.

**Comment by matt-gardner** [comment]
Can you use the plugin stuff that was added a little while ago?  Does that solve the issue?

**Comment by JohnGiorgi** [comment]
It did! Thanks a lot as always. Opened a PR with the fix (#4026).
